src/integrationTest/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/resource/CreateGoalTest.kt

```diff
@@ -0,0 +1,125 @@
+import uk.gov.justice.digital.hmpps.educationandworkplanapi.aValidTokenWithEditAuthority
+import uk.gov.justice.digital.hmpps.educationandworkplanapi.aValidTokenWithViewAuthority
+import uk.gov.justice.digital.hmpps.educationandworkplanapi.app.IntegrationTestBase
+import uk.gov.justice.digital.hmpps.educationandworkplanapi.app.database.jpa.entity.aValidActionPlanEntity
+import uk.gov.justice.digital.hmpps.educationandworkplanapi.app.database.jpa.entity.assertThat
+import uk.gov.justice.digital.hmpps.educationandworkplanapi.bearerToken
+import uk.gov.justice.digital.hmpps.educationandworkplanapi.resource.model.CreateGoalRequest
+import uk.gov.justice.digital.hmpps.educationandworkplanapi.resource.model.ErrorResponse
+import uk.gov.justice.digital.hmpps.educationandworkplanapi.resource.model.aValidCreateGoalRequest
+import uk.gov.justice.digital.hmpps.educationandworkplanapi.resource.model.assertThat
+
+class CreateGoalTest : IntegrationTestBase() {
+
+  companion object {
+    private const val URI_TEMPLATE = "/action-plans/{prisonNumber}/goals"
+  }
+
+  @Test
+  fun `should return unauthorized given no bearer token`() {
+    webTestClient.post()
+      .uri(URI_TEMPLATE, aValidPrisonNumber())
+      .contentType(APPLICATION_JSON)
+      .exchange()
+      .expectStatus()
+      .isUnauthorized
+  }
+
+  // TODO The current implementation returns a 500 instead of a 403. Re-enable this test when the implementation has been fixed
+  @Test
+  @Disabled("The current implementation returns a 500 instead of a 403. Re-enable this test when the implementation has been fixed")
+  fun `should return forbidden given bearer token with view only role`() {
+    webTestClient.post()
+      .uri(URI_TEMPLATE, aValidPrisonNumber())
+      .bearerToken(aValidTokenWithViewAuthority(privateKey = keyPair.private))
+      .contentType(APPLICATION_JSON)
+      .exchange()
+      .expectStatus()
+      .isForbidden
+  }
+
+  @Test
+  fun `should fail to create goal given goal with no steps`() {
+    val prisonNumber = aValidPrisonNumber()
+    val createRequest = aValidCreateGoalRequest(steps = emptyList())
+
+    // When
+    val response = webTestClient.post()
+      .uri(URI_TEMPLATE, prisonNumber)
+      .body(Mono.just(createRequest), CreateGoalRequest::class.java)
+      .bearerToken(aValidTokenWithEditAuthority(privateKey = keyPair.private))
+      .contentType(APPLICATION_JSON)
+      .exchange()
+      .expectStatus()
+      .isBadRequest
+      .returnResult(ErrorResponse::class.java)
+
+    // Then
+    val actual = response.responseBody.blockFirst()
+    assertThat(actual)
+      .hasStatus(BAD_REQUEST.value())
+      .hasUserMessageContaining("At least one Step is required.")
+  }
+
+  // TODO The current implementation returns a 500 instead of a 403. Re-enable this test when the implementation has been fixed
+  @Test
+  @Disabled("The current implementation returns a 500 instead of a 403. Re-enable this test when the implementation has been fixed")
```

Comment:
A NPE is being thrown on line 18 of `UserPrincipalAuditorAware`. The `name` within `AuthAwareAuthenticationToken` is null. Need to investigate...

Reply:
Ah ha, it might be that the JWT we are creating does not have all the required claims (fields) in it. Have a look at `JwtAccessTokenBuilder`, and this function:

```
fun buildAccessToken(
  dpsUsername: String = "auser_gen",
  roles: List<String> = emptyList(),
  privateKey: PrivateKey,
): String =
  Jwts.builder()
    .setSubject(dpsUsername)
    .setClaims(
      mapOf(
        "authorities" to roles,
        "user_name" to dpsUsername,
        "auth_source" to "nomis",
        "user_uuid" to UUID.randomUUID(),
      ),
    )
    .setIssuedAt(Date.from(Instant.now()))
    .setExpiration(Date.from(Instant.now().plus(1, ChronoUnit.HOURS)))
    .signWith(privateKey, SignatureAlgorithm.RS256)
    .compact()
```

compared with what a real DPS JWT token looks like:
```
{
  "sub": "NRUSSELL_GEN",
  "user_name": "NRUSSELL_GEN",
  "auth_source": "nomis",
  "iss": "https://sign-in-dev.hmpps.service.justice.gov.uk/auth/issuer",
  "authorities": [
    "ROLE_PRISON",
    "ROLE_EDUCATION_WORK_PLAN_VIEWER",
    "ROLE_EDUCATION_WORK_PLAN_EDITOR",
    "ROLE_WORK_READINESS_EDIT",
    "ROLE_GLOBAL_SEARCH",
    "ROLE_VIEW_PRISONER_DATA",
    "ROLE_WORK_READINESS_VIEW",
    "ROLE_SLM_SCAN_BARCODE"
  ],
  "client_id": "nathan-russell-1",
  "user_uuid": "9d125413-9222-4bc7-ad08-24c26d64bb90",
  "grant_type": "authorization_code",
  "user_id": "486572",
  "scope": [
    "read",
    "write"
  ],
  "name": "Nathan Russell",
  "exp": 1687781528,
  "jti": "wgOLHml2ISzuRWGIgbRmelB2xYg"
}
```

Reply:
You might also have to debug and step through the Spring Security code in order to determine how it maps the claims in the JWT into a `Authentication` instance (ie. the thing that `UserPrincipalAuditorAware` processes)

Reply:
Doh! You know I can't leave these things alone! 😁 I found the bug, and this is the fix you need to make:
![Screenshot 2023-06-27 at 08 14 20](https://github.com/ministryofjustice/hmpps-education-and-work-plan-api/assets/29573598/80edf9f7-2c0a-438e-beee-53fb0bf4138c)

Basically change line 48 from `setClaims` to `addClaims`

Its because the previous line (`setSubject`) is a convenience method, where `sub` is one of the claims in the JWT. And `setSubject` creates the claims map if it doesn't exist.
And then line 48 comes alone and replaces the claims map with `setClaims` !!!

Reply:
Indeed, `addClaims()` fixed it 👍